Add buttons for dmaker.airfresh.a1/t2017 to xiaomi_miio

Kirmas opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Final step to support dmaker.airfresh.a1/t2017 i added button like:
| Name                     | Comment                  
| -------------------- | -----------------------
| Reset Dust Filter    | Work with both  dmaker.airfresh.a1 and dmaker.airfresh.t2017
| Reset Upper Filter | Work only with dmaker.airfresh.t2017

We can use this workflow and easily add button for other devices (But I have only these). 

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/21857

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

Kirmas committed:
Add buttons for dmaker.airfresh.a1/t2017 to xiaomi_miio
- Reset Dust Filter - for a1 and t2017
- Reset Upper Filter - only for t2017

Kirmas committed:
add Untested file to .coveragerc

Kirmas committed:
Refactoring. Remove unnecessary functions

Kirmas commented on homeassistant/components/xiaomi_miio/button.py:
```diff
@@ -0,0 +1,151 @@
+MODEL_TO_FEATURES_MAP = {
+    MODEL_AIRFRESH_A1: FEATURE_FLAGS_AIRFRESH_A1,
+    MODEL_AIRFRESH_T2017: FEATURE_FLAGS_AIRFRESH_T2017,
+}
```
Was old. I removed.

Kirmas commented on homeassistant/components/xiaomi_miio/button.py:
```diff
@@ -0,0 +1,151 @@
+    async def async_press(self, **kwargs: Any) -> None:
+        """Press the button."""
+        method = getattr(self, self.entity_description.method_press)
+        await method()
```
> Take care and stay safe, @Kirmas!

@rytilahti  Thank you. What you do now to help me finishing this devices integration, it`s very important for me.

> Could we await the `self._try_command()` with the name directly instead of creating those separate coroutines below?
> 
> Plenty of other devices will also have buttons in the future, so it would be better to avoid as much boilerplate as possible.

I refactored code, a little bit. As for me I do some "black python magic", but this is worked, I checked. (Reset the filter at home)

rytilahti replied:
Thanks for working on this, I'll try to convert the vacuum to use this later, and perform clean-ups as I see necessary :-)

rytilahti commented on homeassistant/components/xiaomi_miio/button.py:
```diff
@@ -0,0 +1,136 @@
+        name="Reset Dust Filter",
+        icon="mdi:air-filter",
+        method_press="reset_dust_filter",
+        method_press_error_message="Resetting the dust filter lifetime of the miio device failed",
```
```suggestion
        method_press_error_message="Resetting the dust filter lifetime failed",
```

rytilahti commented on homeassistant/components/xiaomi_miio/button.py:
```diff
@@ -0,0 +1,136 @@
+        name="Reset Upper Filter",
+        icon="mdi:air-filter",
+        method_press="reset_upper_filter",
+        method_press_error_message="Resetting the dust filter lifetime of the miio device failed.",
```
```suggestion
        method_press_error_message="Resetting the dust filter lifetime failed.",
```

rytilahti commented on homeassistant/components/xiaomi_miio/button.py:
```diff
@@ -0,0 +1,136 @@
+    model = config_entry.data[CONF_MODEL]
+    unique_id = config_entry.unique_id
+    device = hass.data[DOMAIN][config_entry.entry_id][KEY_DEVICE]
+    coordinator = hass.data[DOMAIN][config_entry.entry_id][KEY_COORDINATOR]
+    if DATA_KEY not in hass.data:
```
```suggestion
    coordinator = hass.data[DOMAIN][config_entry.entry_id][KEY_COORDINATOR]

    if DATA_KEY not in hass.data:
```

rytilahti commented on homeassistant/components/xiaomi_miio/button.py:
```diff
@@ -0,0 +1,136 @@
+    device = hass.data[DOMAIN][config_entry.entry_id][KEY_DEVICE]
+    coordinator = hass.data[DOMAIN][config_entry.entry_id][KEY_COORDINATOR]
+    if DATA_KEY not in hass.data:
+        hass.data[DATA_KEY] = {}
+    buttons: tuple[str, ...] = ()
```
```suggestion
        hass.data[DATA_KEY] = {}

    buttons: tuple[str, ...] = ()
```

rytilahti commented on homeassistant/components/xiaomi_miio/button.py:
```diff
@@ -0,0 +1,136 @@
+    if model in MODEL_TO_BUTTON_MAP:
+        buttons = MODEL_TO_BUTTON_MAP[model]
+    else:
+        return
+    for description in BUTTON_TYPES:
```
```suggestion
        return

    for description in BUTTON_TYPES:
```

rytilahti replied:
```suggestion
        return

    for description in BUTTON_TYPES:
```

rytilahti commented on homeassistant/components/xiaomi_miio/button.py:
```diff
@@ -0,0 +1,136 @@
+        return
+    for description in BUTTON_TYPES:
+        if description.key not in buttons:
+            continue
+        entities.append(
```
```suggestion
            continue

        entities.append(
```

rytilahti commented on homeassistant/components/xiaomi_miio/button.py:
```diff
@@ -0,0 +1,136 @@
+            and not self.coordinator.data.is_on
+            and not self.entity_description.available_with_device_off
+        ):
+            return False
+        return super().available
```
```suggestion
            return False

        return super().available
```

rytilahti commented on homeassistant/components/xiaomi_miio/button.py:
```diff
@@ -0,0 +1,136 @@
+    ),
+)
+
+AIRFRESH_BUTTON_A1 = (ATTR_RESET_DUST_FILTER,)
+AIRFRESH_BUTTON_T2017 = (
+    ATTR_RESET_DUST_FILTER,
+    ATTR_RESET_UPPER_FILTER,
+)
```
I don't think it's worth creating temporary variables for these, but rather use the values directly inside `MODEL_TO_BUTTON_MAP`.

rytilahti commented on homeassistant/components/xiaomi_miio/button.py:
```diff
@@ -0,0 +1,136 @@
+
+    method_press: str = ""
+    method_press_error_message: str = ""
+    available_with_device_off: bool = True
```
Is this required? If not for now, I'd remove this to keep it simple for now.

rytilahti reviewed (commented):
I'm fine with this after some minor changes:
* Please separate code blocks to logical pieces to make the code easier to read. I added some suggestions there to give an example about this.
* `available_with_device_off` is always True at the moment, this should be removed and re-added later if necessary.

Kirmas committed:
Refactoring:
- separate code blocks to logical pieces to make the code easier to read
-available_with_device_off  was removed
- remove temporary variables

Kirmas commented:
@rytilahti I rechecked the code working and removed the function - so I didn't use the "Commit suggestion" button. But thanks for them.

rytilahti reviewed (approved):
LGTM, as I have never used "buttons" before, I'll add the second-opinion-wanted tag for someone else to chime in. Thanks again! :+1: 

Predicting that someone might ask for missing documentation, my personal opinion is that it does not make sense in this case. There will be many more "buttons" as there exists a wide variety of miio devices, making it futile (and useless) to try to manually document and keep them in sync with the code. If that level of documentation is _really_ wanted, it should be generated automatically from the entity descriptions.

MartinHjelmare commented on homeassistant/components/xiaomi_miio/button.py:
```diff
@@ -0,0 +1,126 @@
+        name="Reset Upper Filter",
+        icon="mdi:air-filter",
+        method_press="reset_upper_filter",
+        method_press_error_message="Resetting the dust filter lifetime failed.",
```
```suggestion
        method_press_error_message="Resetting the upper filter lifetime failed.",
```

MartinHjelmare commented on homeassistant/components/xiaomi_miio/button.py:
```diff
@@ -0,0 +1,126 @@
+    device = hass.data[DOMAIN][config_entry.entry_id][KEY_DEVICE]
+    coordinator = hass.data[DOMAIN][config_entry.entry_id][KEY_COORDINATOR]
+
+    if DATA_KEY not in hass.data:
+        hass.data[DATA_KEY] = {}
+
```
This isn't used.
```suggestion
```

MartinHjelmare commented on homeassistant/components/xiaomi_miio/button.py:
```diff
@@ -0,0 +1,126 @@
+)
+from .device import XiaomiCoordinatedMiioEntity
+
+DATA_KEY = "button.xiaomi_miio"
```
```suggestion
```

MartinHjelmare commented on homeassistant/components/xiaomi_miio/button.py:
```diff
@@ -0,0 +1,126 @@
+        hass.data[DATA_KEY] = {}
+
+    buttons: tuple[str, ...] = ()
+    if model in MODEL_TO_BUTTON_MAP:
+        buttons = MODEL_TO_BUTTON_MAP[model]
+    else:
+        return
```
```suggestion
    if model not in MODEL_TO_BUTTON_MAP:
        return

    buttons = MODEL_TO_BUTTON_MAP[model]
```

MartinHjelmare commented on homeassistant/components/xiaomi_miio/button.py:
```diff
@@ -0,0 +1,126 @@
+    ),
+)
+
+MODEL_TO_BUTTON_MAP = {
```
```suggestion
MODEL_TO_BUTTON_MAP: dict[str, tuple[str, ...]] = {
```

MartinHjelmare commented on homeassistant/components/xiaomi_miio/button.py:
```diff
@@ -0,0 +1,126 @@
+    if DATA_KEY not in hass.data:
+        hass.data[DATA_KEY] = {}
+
+    buttons: tuple[str, ...] = ()
```
```suggestion
```

MartinHjelmare commented on homeassistant/components/xiaomi_miio/button.py:
```diff
@@ -0,0 +1,126 @@
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the button from a config entry."""
+    entities = []
```
Please move the definitions that we don't need until later down after the check of `MODEL_TO_BUTTON_MAP`.

Kirmas committed:
Apply suggestions from code review

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

Kirmas committed:
move the definitions that we don't need until later

MartinHjelmare reviewed (approved):
Great!

Kirmas commented:
@MartinHjelmare thank you for your help.
